src/http-libwebsockets.c

```diff
@@ -96,18 +98,28 @@ typedef struct connection_t {
     struct lws *wsi;
 } connection_t;
 
+// Instantiated for every HTTP request, this holds the statistics to be written in the response
+//
 typedef struct stats_request_state_t {
-    bool callback_completed;
-    bool wsi_deleted;
+    bool callback_completed;  // T: the core has written the global statistics to the stats field
+    bool wsi_deleted;         // T: client has closed, may release this state instance
     qdr_global_stats_t stats;
     qd_http_server_t *server;
     struct lws *wsi;
+    size_t buffer_size;       // extra octets past lws_prefix[LWS_PRE] for HTTP output
+    uint8_t lws_prefix[LWS_PRE];
+    // buffer_size extra octets are appended to this structure when it is allocated. This space is used for the HTTP
+    // response. See new_stats_request_state(), Use &lws_prefix[LWS_PRE] as the start of output buffer.
 } stats_request_state_t;
+static stats_request_state_t *new_stats_request_state(size_t buffer_size);
+static void free_stats_request_state(stats_request_state_t *);
 
+// Context passed to metrics and healthz protocol callbacks. Instantiated by the LWS thread, represents a single HTTP
+// request transaction.
+//
 typedef struct stats_t {
-    size_t current;
-    bool headers_sent;
-    stats_request_state_t *context;
+    stats_request_state_t *state;
```

Comment:
Why not simply move the response_complete flag into `stats_request_state_t` and get rid of `stats_t` ?

Reply:
Most excellent question. It's because I'm paranoid.
It might be possible to get a spurious WRITEABLE event after the request is complete and we've freed the state. Can't tell for sure from the LWS code. Thought best to keep the flag out of the free-able state in case the spurious WRITEABLE occurred after the response was completed (and state freed).